codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/protocols/rpcv2cbor/RpcV2CborProtocolGenerator.kt

```diff
@@ -0,0 +1,90 @@
+import software.amazon.smithy.swift.codegen.integration.HttpBindingResolver
+import software.amazon.smithy.swift.codegen.integration.ProtocolGenerator
+import software.amazon.smithy.swift.codegen.integration.middlewares.ContentLengthMiddleware
+import software.amazon.smithy.swift.codegen.integration.middlewares.ContentTypeMiddleware
+import software.amazon.smithy.swift.codegen.integration.middlewares.OperationInputBodyMiddleware
+import software.amazon.smithy.swift.codegen.model.getTrait
+import software.amazon.smithy.swift.codegen.model.hasTrait
+import software.amazon.smithy.swift.codegen.model.targetOrSelf
+
+class RpcV2CborProtocolGenerator : AWSHTTPBindingProtocolGenerator(RpcV2CborCustomizations()) {
+    override val defaultContentType = "application/cbor"
+    override val protocol: ShapeId = Rpcv2CborTrait.ID
+
+    override val shouldRenderEncodableConformance = true
+
+    override fun getProtocolHttpBindingResolver(ctx: ProtocolGenerator.GenerationContext, defaultContentType: String):
+        HttpBindingResolver = RPCV2CBORHttpBindingResolver(ctx, defaultContentType)
+
+    override fun addProtocolSpecificMiddleware(ctx: ProtocolGenerator.GenerationContext, operation: OperationShape) {
+        super.addProtocolSpecificMiddleware(ctx, operation)
+
+        // Original instance of OperationInputBodyMiddleware checks if there is an HTTP Body, but for AWSJson protocols
+        // we always need to have an InputBodyMiddleware
+        operationMiddleware.removeMiddleware(operation, "OperationInputBodyMiddleware")
+        operationMiddleware.appendMiddleware(operation, OperationInputBodyMiddleware(ctx.model, ctx.symbolProvider, true))
+
+        val hasEventStreamResponse = ctx.model.expectShape(operation.outputShape).hasTrait<StreamingTrait>()
+        val hasEventStreamRequest = ctx.model.expectShape(operation.inputShape).hasTrait<StreamingTrait>()
+
+        // Determine the value of the Accept header
+        val acceptHeaderValue = if (hasEventStreamResponse) {
+            "application/vnd.amazon.eventstream"
+        } else {
+            "application/cbor"
+        }
+
+        // Determine the value of the Content-Type header
+        val contentTypeValue = if (hasEventStreamRequest) {
+            "application/vnd.amazon.eventstream"
+        } else {
+            defaultContentType
+        }
+
+        // Middleware to set smithy-protocol and Accept headers
+        // Every request for the rpcv2Cbor protocol MUST contain a smithy-protocol header with the value of rpc-v2-cbor
+        val smithyProtocolRequestHeaderMiddleware = MutateHeadersMiddleware(
+            overrideHeaders = mapOf(
+                "smithy-protocol" to "rpc-v2-cbor",
+                "Accept" to acceptHeaderValue
+            )
+        )
+
+        operationMiddleware.appendMiddleware(operation, smithyProtocolRequestHeaderMiddleware)
+        operationMiddleware.appendMiddleware(operation, CborValidateResponseHeaderMiddleware)
```

Comment:
I think it would be great to have general purpose middleware for verifying headers instead of RPCv2 CBOR specific middleware. Similar to how the `MutateHeadersMiddleware` allows passing in multiple headers to add to request, there could be a middleware called something along the lines of `VerifyHeadersMiddleware` that verifies passed in headers; you could pass in `smithy-protocol : rpc-v2-cbor` to it to serve the specific purpose here.

Reply:
This is a great thought but I'm not sure this pattern is very generalizable. In this case we want to ensure we're validating a header **prior to deserialization** and we don't want anything else but this one specific header to be deserialized before continuing, even to check the other headers. It also gets tricky when there may be more than one correct value for a header with custom error messages, leading to a lot of function input being generated into our clients.

Reply:
By the time `readBeforeDeserialization` interceptor hook is reached in operation flow, unless I'm mistaken, all response headers are already deserialized; i.e., headers are already deserialized into `context.getResponse().header`. I think what's not deserialized yet is the payload.

Reason for suggestion was for future extensibility; off top of my head, other place in our codebase where we validate on response headers is for flexible checksum, but I think that one should stay where it is in its own checksum middleware. 

If generic method of validating on response header is set up as part of RPCv2 CBOR, then future extension would be easy. Otherwise, someone would need to look into RPCv2 CBOR SEP to refactor existing code; more likely, just add another specialized middleware for future use case, not knowing this should probably be refactored.

It depends on how big of a work it would be though; if it's more than just quick refactoring, I'd be fine with leaving this out of this task for now.

Reply:
Discussed offline; and we concluded that keeping this header presence check specific to RPCv2 CBOR middleware makes sense, given checking header presence just to determine go/no-go for body deserialization is probably going to be unique to RPCv2 CBOR flow.